src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -81,23 +74,42 @@ void remove_virtual_functionst::remove_virtual_function(
   const code_function_callt &code=
     to_code_function_call(target->code);
 
-  const auto &vcall_source_loc=target->source_location;
-
   const exprt &function=code.function();
-  assert(function.id()==ID_virtual_function);
-  assert(!code.arguments().empty());
-
-  functionst functions;
+  INVARIANT(
+    function.id()==ID_virtual_function,
+    "remove_virtual_function must take a virtual function call instruction");
+  INVARIANT(
+    !code.arguments().empty(),
+    "virtual function calls must have at least a this-argument");
+
+  dispatch_table_entriest functions;
   get_functions(function, functions);
 
+  remove_virtual_function(
+    goto_program,
+    target,
+    functions,
+    virtual_dispatch_fallback_actiont::CALL_LAST_FUNCTION);
+}
+
+void remove_virtual_functionst::remove_virtual_function(
+  goto_programt &goto_program,
+  goto_programt::targett target,
+  const dispatch_table_entriest &functions,
+  virtual_dispatch_fallback_actiont fallback_action)
+{
+  const code_function_callt &code=
+    to_code_function_call(target->code);
+
   if(functions.empty())
   {
     target->make_skip();
     return; // give up
   }
 
   // only one option?
-  if(functions.size()==1)
+  if(functions.size()==1 &&
+     fallback_action==virtual_dispatch_fallback_actiont::CALL_LAST_FUNCTION)
   {
     assert(target->is_function_call());
```

Comment:
It may be worthwhile moving this assertion to the beginning of this (new) function.